src/lib.rs

```diff
@@ -0,0 +1,718 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+use borsh::schema::{BorshSchemaContainer, Declaration, Definition, Fields, VariantName};
+use schemars::schema::{RootSchema, Schema};
+use serde::{Deserialize, Serialize};
+
+/// Current version of the ABI schema format.
+const ABI_SCHEMA_SEMVER: &str = env!("CARGO_PKG_VERSION");
+
+/// Contract ABI.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct AbiRoot {
+    /// Semver of the ABI schema format.
+    abi_schema_version: String,
+    /// Metadata information about the contract.
+    pub metadata: AbiMetadata,
+    /// Core ABI information (functions and types).
+    pub abi: AbiEntry,
+}
+
+impl AbiRoot {
+    #[cfg(feature = "versioned-entries")]
+    pub fn new(metadata: AbiMetadata, versioned_abi: VersionedAbiEntry) -> AbiRoot {
+        AbiRoot {
+            abi_schema_version: versioned_abi.abi_schema_version,
+            metadata,
+            abi: versioned_abi.abi,
+        }
+    }
+
+    pub fn schema_version(&self) -> &str {
+        &self.abi_schema_version
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Default)]
+pub struct AbiMetadata {
+    /// The name of the smart contract.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    /// The version of the smart contract.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub version: Option<String>,
+    /// The authors of the smart contract.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub authors: Vec<String>,
+    /// Other arbitrary metadata.
+    #[serde(default, flatten, skip_serializing_if = "HashMap::is_empty")]
+    pub other: HashMap<String, String>,
+}
+
+/// Core ABI information.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct AbiEntry {
+    /// ABIs of all contract's functions.
+    pub functions: Vec<AbiFunction>,
+    /// Root JSON Schema containing all types referenced in the functions.
+    pub root_schema: RootSchema,
+}
+
+/// Core ABI information, with schema version.
+#[cfg(feature = "versioned-entries")]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct VersionedAbiEntry {
+    /// Semver of the ABI schema format.
+    abi_schema_version: String,
+    #[serde(flatten)]
+    pub abi: AbiEntry,
+}
```

Comment:
I like the idea of statically known versions, but I wonder if there is a better way to represent this. Have you thought about something like this (very roughly outlined)?

```rust
pub enum VersionedAbiEntry {
    #[serde(rename = "0.1")]
    V0_1(AbiEntryDeprecated),
    #[serde(rename = "0.2")]
    V0_2(AbiEntryDeprecated),
    #[serde(rename = "1.0")]
    V1_0(AbiEntry),
}
```

And then change `abi.json` to look like this:

```json
{
  "metadata": {
    "name": "abi",
    "version": "0.1.0",
    "authors": [
      "Near Inc <hello@nearprotocol.com>"
    ]
  },
  "1.0": {
    ... // AbiEntry contents
  }
}
```

I never liked that `abi.json` has a field named `abi`, so maybe it is time to change it? :)

Reply:
<div align=right>

_What's this? ... Okay I like it Picasso._[ <sup>1</sup> ](https://www.youtube.com/watch?v=EyKF_cOuwno)

</div>

> I never liked that abi.json has a field named abi, so maybe it is time to change it? :)

Or `abi` inside `AbiRoot`, lol.

Yeah, I really like this idea.

Reply:
The only thing I'd think about this, is that supporting the different types might be awkward to work with. Seems like the version should be based on the whole schema, rather than just the ABI section, maybe. Would match JSON schema more closely. Supporting this would be straightforward in Rust, but unclear how this would work well with JS/TS or other languages without enums like this